### PR TITLE
1st commit of fusion capable decoder, json support

### PR DESCRIPTION
**Pull request by jeffnye-gh**
fusion decoder implementation. This is suitable for analysis.

Adds fusion parameters to fusion.yaml. And specifies external file definition of fusion groups using JSON

New statistics for fused instructions and fusion group stats

HCache as separate structure in fusion/fusion

Includes the FSL.md domain language description

Inst.hpp modified to add FUSED and FUSION_GHOST status as extended status

Support for non-sequential program ID. Each instr has a PID increment value, and methods

Formatting, clean up of Inst.hpp

JSON support added to Fusion.hpp

New exceptions for JSON errors

Fusion default transform no longer modifies the input buffer

HCache testbench functions

### CODE REVIEW

File: CMakeLists.txt
Comment (klingaard) [inline CMakeLists.txt:63]: The reason I had `mavis` as a requirement for the include directive (`#include "mavis/Mavis.h"`) was to prevent header file naming collisions if they happened.  I don't have complete control on Mavis, so if someone added `mavis/CommonTypes.hpp` and we had `CommonTypes.hpp` the compiler wouldn't get confused.   Can this rule be put back?

File: arches/fusion.yaml
Comment (klingaard) [inline arches/fusion.yaml:22]: Should be able to say:
```
include: big_core.yaml
```
and remove everything from here down.

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:1]: Can you make the private function names match the coding style guide?  Specifically, missing a trailing `_`

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:95]: Considering using `std::numeric_limits<uint32_t>::max()`

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:243]: Nit: missing trailing `_`

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:298]: Naked pointer.  `std::unique_ptr<FusionType> fuser_;`

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:304]: Tabbing is funny...

Is this const expression?  Non-const static variables in a perf model that can support heterogeneous cores (an MC situation) might be problematic.

File: core/Decode.hpp
Comment (klingaard) [inline core/Decode.hpp:306]: Nit: naming doesn't match the rest of the code.

File: core/Decode.cpp
Comment (klingaard) [inline core/Decode.cpp:142]: nit: tabbing is off

File: core/Decode.cpp
Comment (klingaard) [inline core/Decode.cpp:158]: Nit: please add `{}` around the `if` statement body.

File: core/Decode.cpp
Comment (klingaard) [inline core/Decode.cpp:184]: nit: tabbing funny.

Might want to throw this behind a parameter like `fusion_debug`

File: core/FusionDecode.cpp
Comment (klingaard) [inline core/FusionDecode.cpp:63]: Suggest another variable name other than `m`.  If I wanted to search for the use of `m`, I get a LOT of hits.

File: core/FusionDecode.cpp
Comment (klingaard) [inline core/FusionDecode.cpp:85]: `++q`?  :) 

You probably should ensure that `q != insts.end()` unless that's not possible

File: core/FusionDecode.cpp
Comment (klingaard) [inline core/FusionDecode.cpp:72]: Suggest another variable name other than `q`

File: core/FusionDecode.cpp
Comment (klingaard) [inline core/FusionDecode.cpp:124]: Nit: tabbing funny.

**Review by klingaard — COMMENTED** [review]
More comments to come, but this great!

File: CMakeLists.txt
Comment (jeffnye-gh) [reply CMakeLists.txt:63]: Yes, of course.

I'll also make sure the fusion headers are consistent with this in olympia. 
e.g. #include "fusion/FusionWhatsIt.hpp"

It will look like this now:

# Include directories
include_directories (core mss sim)
include_directories (SYSTEM fusion)
include_directories (SYSTEM mavis)
include_directories (SYSTEM stf_lib)

File: arches/fusion.yaml
Comment (jeffnye-gh) [reply arches/fusion.yaml:22]: Corrected,

I do not know what I was doing before but it was not picking up fusion_enable state. Probably indentation issue. Fixed now

File: core/Decode.cpp
Comment (jeffnye-gh) [reply core/Decode.cpp:142]: I will run clang-format on Decode.cpp and .hpp. That will fix this. 

There will likely be unrelated format changes.

File: core/Decode.cpp
Comment (jeffnye-gh) [reply core/Decode.cpp:158]: clang format did not correct this. I've fixed it manually.

File: core/Decode.cpp
Comment (jeffnye-gh) [reply core/Decode.cpp:184]: tabbing fixed with clang-format.

I have added the parameter.

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:1]: Yes, :)

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:95]: done

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:243]: yes, fixing all these.

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:95]: change made

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:298]: fixed

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:304]: clang format now run on this file. 

fusionGroupCfgs removed, no longer needed when JSON support added

File: core/Decode.hpp
Comment (jeffnye-gh) [reply core/Decode.hpp:306]: fixed

File: core/FusionDecode.cpp
Comment (jeffnye-gh) [reply core/FusionDecode.cpp:63]: i fixed this one, the q/a/b/m.

File: core/FusionDecode.cpp
Comment (jeffnye-gh) [reply core/FusionDecode.cpp:72]: fixed

File: core/FusionDecode.cpp
Comment (jeffnye-gh) [reply core/FusionDecode.cpp:85]: advance was a carry over from before it was known to always be a constant increment.

fixed

it is not possible to exceed the range but it is not obvious from this function,  so I added an assert.

File: core/FusionDecode.cpp
Comment (jeffnye-gh) [reply core/FusionDecode.cpp:124]: fixed w/ clang-format

File: fusion/FSL.md
Comment (arupc) [inline fusion/FSL.md:625]: Using UID style specification will have an impact of readability and may need comments describing the UID.  Is there a reason why we need the UID style specification and not just use the abstract instruction style specification?

File: fusion/FSL.md
Comment (arupc) [inline fusion/FSL.md:715]: Is there a way to fuse one or more classes of instruction with other instructions? For example, is there a way to define a `logical immediate` class of instructions for `andi`, `ori`, `xori` and define a sequence in terms of that class, instead of individual instructions?

File: fusion/FSL.md
Comment (arupc) [inline fusion/FSL.md:152]: Trying to understand the motivation behind allowing in-place modification. Can we have an example of where this will be preferred / useful?

File: fusion/FSL.md
Comment (arupc) [inline fusion/FSL.md:479]: Can prolog elements be considered as a form of constraint? If so, can this be made part of constraints?

File: fusion/FSL.md
Comment (arupc) [inline fusion/FSL.md:199]: If we do not anticipate having more than one sequence, constraint and conversion clause per transform, wondering if we need to specify the arguments `iop1`, `seq1`, `cons1` in the agument lists for every transform. Instead, is it okay to make them implicit?

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:118]: To modify the input buffer directly rather than modify and copy to a new buffer.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:129]: A clause is a general term for a syntax construct, it is a noun, phases are the operations, verbs. We can discuss more but I like the convenience of this correlation to natural language for explaining things.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:152]: To minimize copies. e.g. take the case where there are no successful transforms performed, without in-place, an unmodified sequence of ops needs to be copied to the output.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:199]: Yes, it's a question of language design/style. Within the constraints you mention there is no technical need for the arguments.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:218]: yes, that is a typo.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:412]: This is just preference when implementing the parser/grammar, 

Use of '#' for preprocessor directives dates to K&R supposedly selected because it was not prevalent in any languages at the time. but now #'s are found in some assembly languages. like ARM asm.

At the moment '#' is the current comment character. But this may change.

I chose the verilog syntax because I'm used to it and it's well known and back ticks are not at all common in the assembly languages I have been exposed too.

But again, boils down to my preference.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:479]: I think of them more as context although that distinction is not always sharp.  Having them where they are in the syntax makes their scope explicit, they are used by the other clauses, it also makes the library style a consistent extension of this style, the prolog elements can be referenced as entities rather than a sub-entity of a named constraint object.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:625]: The UID form lends it's self to automated flows. The UID form comes out of the practical work I'm doing. If I do not need to express operand specific constraints then I can more fully automate the discovery of possible fusion tuples, express these to the model, perform the simulation and extract the results. This is a very low touch methodology.

Having said that UIDs is not the only method. There is support for assembly language expression. "...as a set of assembly language instructions with..."

Comments are possible, there are no comments per UID in this example since the associated instruction examples are shown directly above, line ~610

Once the FSL parser has an integrated RISC-V assembly language parser that supports abstract operands the assembly language form of these examples will be available.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:650]: Yes. The _pass_ keyword assists this.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:715]: There is no way to do this in the current grammar. Let me give it some thought.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:775]: Yes.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:797]: For confirming the type assignment and syntactic correctness of it's usage. 

I have a named operand Q1. I may not care what value is referred to by Q1, but I may care that it is a GPR. And by knowing it's type I can also tell if the assembly language statement using Q1 is correctly formed.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:910]: yes, sounds reasonable.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:1066]: For debug purposes sure.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:1085]: Yes.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:715]: I think the proper way to handle the need is to provide syntax for composition of sequences from other sequences rather than rely on meta-data based classifications. 

I'll do this in a future version.

File: fusion/FSL.md
Comment (jeffnye-gh) [reply fusion/FSL.md:199]: For now I'm going to leave the arguments as required. I need to see a parser handling scope properly. Once I have proven I can handle scope for large/complex translation units I'll revisit this. It is marked as part of the TODO section.

File: fusion/FSL.md
Comment (arupc) [reply fusion/FSL.md:726]: Can we please capture this feature/need in an issue? Or if you would like, may be create one issue to capture various future improvements of FSL?

**Review by arupc — APPROVED** [review]

**Comment by arupc** [comment]
This looks good to me. Please feel to merge this, unless Knute has any other comment.

This PR is a significant contribution. There may be opportunity for wider acceptance and use of this language, beyond this SIG and Olympia. After we merge this PR, would you be interested in doing a RISCV-style public review of the FSL.md document, as a means of gathering feedback outside of the SIG for future improvements?

**Comment by jeffnye-gh** [comment]
I have created the issue which captures the enhancements for composition and argument lists. [#165](https://github.com/riscv-software-src/riscv-perf-model/issues/165)

I can not do the merge, its requires Knute or other admin for the repo.

Re:  "After we merge this PR, would you be interested in doing a RISCV-style public review of the FSL.md"

Sure, I would create a more or less standard BNF expression of the grammar based on this review feedback, and then public review would be good.

**Pull request merged**
